lib/predicate/registry.js

```diff
@@ -0,0 +1,353 @@
+        });
+        return true;
+    },
+    log(v) {
+        if (typeof v !== 'object') {
+            return false;
+        }
+        return ['trace', 'info'].every(k => typeof v[k] === 'function');
+    },
+};
+
+function createKey(eventInfo) {
+    return `${eventInfo.eventName}|${eventInfo.bucket}`;
+}
+
+const s3ObjectMapping = {
+    key(v) {
+        if (this.objectKey !== v) {
+            this.objectKey = v;
+        }
+    },
+    contentType(v) {
+        if (this.headers['content-type'] !== v) {
+            this.headers['content-type'] = v;
+        }
+    },
+    metadata(v) {
+        Object.keys(v).forEach(k => {
+            const tag = k.toLowerCase();
+            this.headers[`x-amz-meta-${tag}`] = v[k];
+        });
+    },
+};
+
+function updateEventInfo(event, output) {
+    const s3Obj = event.Records[0].s3.object;
+    Object.keys(s3ObjectMapping).forEach(k => {
+        s3ObjectMapping[k].call(output, s3Obj[k]);
+    });
+    return output;
+}
+
+// TODO: improve field extraction
+function makePredicateEvent(runParams, body) {
+    return {
+        Records: [{
+            eventVersion: '2.0',
+            eventSource: 'aws:s3',
+            awsRegion: runParams.region,
+            eventTime: new Date(),
+            eventName: runParams.eventName,
+            userIdentity: {
+                principalId: runParams.userId,
+            },
+            requestParameters: {
+                sourceIPAddress: runParams.sourceIPAddress,
+            },
+            responseElements: {
+                'x-amz-request-id': runParams.amzRequestId,
+                'x-amz-id-2': runParams.amzRequestId2,
+            },
+            s3: {
+                s3SchemaVersion: '1.0',
+                configurationId: '00000000-0000-0000-0000-000000000000',
+                bucket: {
+                    name: runParams.bucket,
+                    ownerIdentity: {
+                        principalId: runParams.bucketOwnerId,
+                    },
+                    arn: runParams.bucketArn,
+                },
+                object: {
+                    key: runParams.key,
+                    size: runParams.size,
+                    eTag: runParams.etag,
+                    contentType: runParams.contentType,
+                    metadata: runParams.metadata,
+                    body,
+                    versionId: 'N/A',
+                    sequencer: 'N/A',
+                },
+            },
+        }],
+    };
+}
+
+function tryRequire(p) {
+    const out = {};
+    try {
+        const fnpath = path.resolve(process.cwd(), p);
+        const fn = require(fnpath);
+        if (typeof fn !== 'function') {
+            throw new TypeError('Required object not function');
+        }
+        out.fn = fn;
+    } catch (e) {
+        out.err = e;
+    }
+    return out;
+}
+
+function getPredicate(registry, runParams, callback) {
+    const errs =
+        validate(runParamsSpec, runParams, runParams);
```

Comment:
I understand why this would be confusing... It's just saying that we want to validate `runParams` against `runParamsSpec` and we want every `runParamsSpec` member function to be run in the context of `runParams`, letting us do [this](https://github.com/austo/S3/blob/master/lib/predicate/registry.js#L32).